src/docbuilder/metadata/lib.rs

```diff
@@ -0,0 +1,631 @@
+#![deny(missing_docs)]
+
+// N.B. requires nightly rustdoc to document until intra-doc links are stabilized.
+//! Collect information that allows you to build a crate the same way that docs.rs would.
+//!
+//! This library is intended for use in docs.rs and crater, but might be helpful to others.
+//! Here is an example use of the crate:
+//!
+//! ```
+//! # fn main() -> Result<(), Box<dyn std::error::Error>> {
+//! use std::process::Command;
+//! use metadata::Metadata;
+//!
+//! // First, we need to parse Cargo.toml.
+//! let source_root = env!("CARGO_MANIFEST_DIR");
+//! let metadata = Metadata::from_crate_root(&source_root)?;
+//!
+//! // Next, learn what arguments we need to pass to `cargo`.
+//! let targets = metadata.targets();
+//! let mut cargo_args = metadata.cargo_args();
+//! cargo_args.push(targets.default_target.into());
+//!
+//! // Now, set up the `Command`
+//! let mut cmd = Command::new("cargo");
+//! cmd.args(cargo_args);
+//! for (key, value) in metadata.environment_variables() {
+//!     cmd.env(key, value);
+//! }
+//!
+//! // Finally, run `cargo doc` on the directory.
+//! let result = cmd.output()?;
+//! # Ok(())
+//! # }
+//! ```
+
+use std::collections::{HashSet, HashMap};
+use std::io;
+use std::path::Path;
+
+use thiserror::Error;
+use toml::{map::Map, Value};
+
+/// The target that `metadata` is being built for.
+///
+/// This is directly passed on from the Cargo [`TARGET`] variable.
+///
+/// [`TARGET`]: https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables-cargo-sets-for-build-scripts
+pub const HOST_TARGET: &str = env!("DOCS_RS_METADATA_HOST_TARGET");
+/// The targets that are built if no `targets` section is specified.
+///
+/// Currently, this is guaranteed to have only [tier one] targets.
+/// However, it may not contain all tier one targets.
+///
+/// [tier one]: https://doc.rust-lang.org/nightly/rustc/platform-support.html#tier-1
+pub const DEFAULT_TARGETS: &[&str] = &[
+    "i686-pc-windows-msvc",
+    "i686-unknown-linux-gnu",
+    "x86_64-apple-darwin",
+    "x86_64-pc-windows-msvc",
+    "x86_64-unknown-linux-gnu",
+];
+
+/// The possible errors for [`Metadata::from_crate_root`].
+#[derive(Debug, Error)]
+pub enum MetadataError {
```

Comment:
`#[non_exhaustive]`?

Reply:
Is it non-exhaustive? I don't see this ever expanding in the future. If for some reason it does we can always make a major version bump.

Reply:
I'd prefer to avoid making major version bumps if we can avoid id.

Reply:
Added